### PR TITLE
Enable native AArch64 Ubuntu CI jobs for the JIT

### PR DESCRIPTION
**Pull request by diegorusso**
Enable native AArch64 Ubuntu CI jobs for the JIT. In order to achieve that I had to refactor the jit.yml workflow:
* Split jobs by OS (Windows, Mac, Linux)
* LLVM version is now a workflow variable accessible by all jobs
* Ubuntu updated from 22.04 to 24.04
* Run  the pipeline only when the branch is `main` or `3.*` like we have in `build.yaml`

The refactoring was needed because these jobs used different fields from the earlier unified matrix. By splitting it, we can better tailor what fields to use for every OS.
Also this allows to exclude AArch64 Ubuntu jobs on forks. Earlier this wasn't possible because the workflow was using in the matrix definition an `include` that is always executed after an `exclude`. There was no way to exclude AArch64 Linux runners for forks.
As a consequence, the job names are not re
porting anymore the target/compiler but more generically: `OS (arch or runner, Debug/Release)` like the image below:

Screenshot 2024-11-29 at 14 28 07

### CODE REVIEW

**Comment by diegorusso** [comment]
This is a continuation of https://github.com/python/cpython/pull/125786

**Comment by hugovk** [comment]
Exciting!

Could we add something like `is-fork` from https://github.com/python/cpython/pull/121809/files to only include these custom runners for upstream CPython runs and not for any runs on forks?

File: .github/workflows/build.yml
Comment (diegorusso) [inline .github/workflows/build.yml]:
```
os:
- ubuntu-24.04
- ubuntu-24.04-aarch64
  is-fork: false
```

**Comment by diegorusso** [comment]
> Exciting!
> 
> Could we add something like `is-fork` from https://github.com/python/cpython/pull/121809/files to only include these custom runners for upstream CPython runs and not for any runs on forks?

Thanks @hugovk for the help on Discord DM :) I think we have now what you were asking.

File: .github/workflows/jit.yml
Comment (savannahostrowski) [inline .github/workflows/jit.yml:129]: In one way, I might suggest we keep the emulated job for those running CI on forks, but in another way, I have never actually run into a problem with emulated Linux runs that didn't require me to add to our skipped tests. So...I think I'm probably okay with removing this and just accepting there's no coverage in the forked repo case.

That said, we can also remove our skipped tests file if we are removing this job https://github.com/python/cpython/blob/main/Tools/jit/ignore-tests-emulated-linux.txt

File: .github/workflows/jit.yml
Comment (savannahostrowski) [inline .github/workflows/jit.yml:84]: I wonder if it's worth trying to bump all jobs to 24.04 at the same time. I downgraded us a couple of weeks back because there were issues with the runner images, but maybe we could see if that's resolved now.

File: .github/workflows/jit.yml
Comment (savannahostrowski) [inline .github/workflows/jit.yml]: Is this flipped? Don't we only want to run these jobs if the org is python?

**Review by savannahostrowski — CHANGES_REQUESTED** [review]

File: .github/workflows/jit.yml
Comment (diegorusso) [reply .github/workflows/jit.yml:84]: Done!

File: .github/workflows/jit.yml
Comment (diegorusso) [reply .github/workflows/jit.yml]: yes, copy/paste and forgot about it :)

**Comment by brandtbucher** [comment]
This is probably going to conflict with GH-127212. I think maybe we should merge that one first (resolving the conflict shouldn't be too bad, since you're just removing the conflicting lines anyways).

**Review by brandtbucher — DISMISSED** [review]
Happy with the JIT changes, thanks!

File: .github/workflows/jit.yml
Comment (diegorusso) [reply .github/workflows/jit.yml:129]: Done!

**Comment by diegorusso** [comment]
I have made the requested changes; please review again.

**Review by savannahostrowski — DISMISSED** [review]
LGTM for JIT CI updates! Thanks Diego!

**Comment by hugovk** [comment]
I pushed this branch to my fork and the aarch64 ones are just sitting waiting for a runner:
https://github.com/hugovk/cpython/actions/runs/12055728919

And they won't find one, because the custom runners are only available for upstream.

So we'll need to do something like the `no-fork` trick for `jit.yml`, like we did for the ubuntu ones in `build.yml` (copied from the macOS ones that run on `ghcr.io/cirruslabs/macos-runner:sonoma` for upstream and fallback to `macos-14` for forks).

The `ubuntu-24.04-aarch64` ones should only run in upstream. We should decide if, for forks, whether there should be a fallback to GH's own `ubuntu-24.04` runners, or if we don't need these jobs for forks at all.

**Comment by savannahostrowski** [comment]
Copying this from a conversation on Discord for posterity.

Thinking about this more, I think there may be a missing security boundary here. While the logic to check if the PR is on a fork may limit runs, I'm pretty sure that this is still a security risk because someone could open a PR to their fork, which removes the logic that checks if it's a fork, and then the custom runners would be used.

Unless I'm missing something, I think we may need to think through this a bit more? GitHub also recommends that self-hosted runners only be used on private repos.

https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners#self-hosted-runner-security

**Comment by diegorusso** [comment]
Hello, I reworked the PR "completely". @savannahostrowski @brandtbucher please have a look.

**Review by hugovk — APPROVED** [review]
Thank you!

File: .github/workflows/jit.yml
Comment (savannahostrowski) [inline .github/workflows/jit.yml:149]: Nit, but can we use all lowercase for LLVM? The line above and the env var use upper case.

**Review by savannahostrowski — APPROVED** [review]
One nit but, otherwise, this LGTM! Thanks for fighting with this, Diego!

File: .github/workflows/jit.yml
Comment (savannahostrowski) [inline .github/workflows/jit.yml:111]: Actually, is there any way that we can more explicitly say that the macos-14 jobs are M1/arm jobs? Just to make it more obvious to folks less familiar.

File: .github/workflows/jit.yml
Comment (diegorusso) [reply .github/workflows/jit.yml:149]: Sure, I'll put everything lowercase.

File: .github/workflows/jit.yml
Comment (diegorusso) [reply .github/workflows/jit.yml:111]: yes, there is :) I've just pushed a new commit that specifies `aarch64` or `x86_64` for both Linux and macOS so names are more consistent

Comment (diegorusso) [reply .github/workflows/jit.yml:111]: Screenshot 2024-12-02 at 17 09 35

**Review by savannahostrowski — APPROVED** [review]
Thanks Diego! This looks great to me!

**Comment by savannahostrowski** [comment]
@brandtbucher If you're okay with these changes, I'd be ready to merge this.

**Comment by brandtbucher** [comment]
So, it looks like it works correctly (thanks for figuring it out). I'm just bummed because I find the new structure quite a bit harder to follow and work with than before.

My understanding is that the restructuring is necessary because we can't exclude jobs from a matrix that's built up of includes, since excludes runs before includes.

I'd like to explore the options a bit further before landing this, since GHA has lots of different, uh, "features" for this sort of logic. For instance: could we turn [this line](https://github.com/python/cpython/blob/fc5a0dc22483a35068888e828c65796d7a792c14/.github/workflows/jit.yml#L121) into `runner.os == 'Linux' && (matrix.architecture == 'x86_64' || !is-fork)` (or whatever the correct syntax is)? Or is that too late in the process?

**Comment by brandtbucher** [comment]
See GH-127581 for an alternative approach. The two AArch64 Linux jobs [run as expected](https://github.com/python/cpython/actions/runs/12149020548/job/33879127581) on CPython org CI, and [complete immediately](https://github.com/brandtbucher/cpython/actions/runs/12148883257/job/33878723244) on the fork's CI.

When the runners (presumably) become free for all, we can just remove the guard on the "Native Linux" job.

**Comment by diegorusso** [comment]
We tried at some point to put the `if` in the job but it doesn't work. The reason is very simple: in order to run the job, it requires a runner, but on forks the aarch64 runners will never be available because they are not free ([that's the same reason we exclude Cirrus macOS from forks](https://github.com/python/cpython/blob/main/.github/workflows/build.yml#L215))
So the line you suggested is executed only on runners which are assigned to the job. To show you what it is happening, I've removed the exclude and added that `if` (that will be never executed anyway) on a branch on my fork: https://github.com/diegorusso/cpython/commit/9a1130bbbcacaebc3ba330bfeeb40c7413e669a9

As you can see from the screenshot below, the aarch64 jobs will be pending forever because the matrix includes them but they will never be allocated on forks. The only way possible is to exclude them from the matrix altogether with `exclude`.

Screenshot 2024-12-03 at 21 52 34

The only way possible is to exclude them when building the matrix. Unfortunately as the `include` is run after the `exclude`, if we have both, it will schedule just the release aarch64 job on forks (thanks to @hugovk to figure this out) because by default debug is not added hence it is false.

When reading the fields in the include, I realized that:
* target is used just in names of the job
* architecture is used just for windows builds. The architecture can be extracted in another way
* the runner can be stripped if we split these jobs

Hence I decided to do the refactoring. As further improvement, this could lead to an additional refactoring by moving the build logic in the `reusable-*.yml` definitions. So we have one place where the build logic is contained.

I'm open to other solutions (if there are)

**Comment by diegorusso** [comment]
After messaging with @brandtbucher  on Discord, I decided to close this PR as leave it for history.
I'll open a new PR that implements what Brandt has suggested here: https://github.com/python/cpython/pull/127581

Moreover:
*  we don't need the whole refactoring
* LLVM version can stay in the matrix
* the exclude in the build.yml is OK
* we don't need to filter JIT workflow for specific branches. We actually want the JIT workflow to run on every branch
* It's OK to update the ubuntu version to 24.04

**Comment by diegorusso** [comment]
This is the new PR: https://github.com/python/cpython/pull/127584

**Pull request closed without merging**
